### PR TITLE
Dev ipa+isntantid

### PR DESCRIPTION
**Pull request by ccssu**
fix: https://github.com/siliconflow/onediff/issues/795

## Doc

- [ ] join ci
- [x]  [Accelerating ComfyUI_InstantID with OneDiff](https://github.com/siliconflow/onediff/tree/dev_ipa%2Bisntantid/onediff_comfy_nodes/modules/oneflow/hijack_comfyui_instantid#readme)
- [x] [Accelerating ComfyUI_IPAdapter_plus with OneDiff](https://github.com/siliconflow/onediff/blob/3d5010e7be267c2f1cd10e413285f1e07d7b74a8/onediff_comfy_nodes/modules/oneflow/hijack_ipadapter_plus/README.md)
- [ ] [Accelerating ComfyUI_IPAdapter_plus + ComfyUI_InstantID with OneDiff]()
### Reference installation
```shell
# 1. onediff
git clone https://github.com/siliconflow/onediff.git
cd onediff 
git checkout dev_ipa+isntantid
pip install -e .
# 2. onediff_comfy_nodes
ln -s $(pwd)/onediff_comfy_nodes path/to/ComfyUI/custom_nodes/
# or
#. cp -r onediff_comfy_nodes path/to/ComfyUI/custom_nodes/
```

### Test version
```
ComfyUI:
  commit 4bd7d55b9028d79829a645edfe8259f7b7a049c0 (HEAD -> master, origin/master, origin/HEAD)
  Author: comfyanonymous <comfyanonymous@protonmail.com>
  Date:   Thu Apr 11 22:43:05 2024 -0400

github: https://github.com/cubiq/ComfyUI_InstantID
ComfyUI_InstantID$ git log
  commit e9cc7597b2a7cd441065418a975a2de4aa2450df (HEAD -> main, origin/main, origin/HEAD)
  Author: matt3o <matt3o@gmail.com>
  Date:   Tue Apr 9 14:05:15 2024 +0200

ComfyUI_IPAdapter_plus$ git log
commit 417d806e7a2153c98613e86407c1941b2b348e88 (HEAD -> main, origin/main, origin/HEAD)
Author: matt3o <matt3o@gmail.com>
Date:   Wed Apr 10 13:28:41 2024 +0200

ComfyUI-AnimateDiff-Evolved$ git log
commit f9e0343f4c4606ee6365a9af4a7e16118f1c45e1 (HEAD -> main, origin/main, origin/HEAD)
Merge: f7138b2 93af734
Author: Jedrzej Kosinski <kosinkadink1@gmail.com>
Date:   Sat Apr 6 17:32:15 2024 -0500


```

### test
You can Load these images in ComfyUI to get the full workflow.

#### ipa
ipadapter

#### isntantid
comfyui_instantid_00

#### ipa+isntantid
ipa+isntantid

### CODE REVIEW

**Comment by Gutianpei** [comment]
I'm having two issues with this branch when compiling IPAdapter and Animatediff together:
1. Changing reference image of ipadapter is not working
    - this can be solved by remove deepcopy in animatediff comfy node but the solution is not ideal
2. Multiple ipadapter is not working
    - only the first reference image will be embed into UNet, if I remove the deepcopy in animatediff then only the last image will be referenced.

Please kindly look into these issues for future development, as ipadapter+animatediff have a great potential recently.

**Comment by ccssu** [comment]
> I'm having two issues with this branch when compiling IPAdapter and Animatediff together:

Fixed the use of AnimateDiff combined with IPA in the same way as before. Please update the code accordingly。
### Reference installation
```shell
# onediff
git clone https://github.com/siliconflow/onediff.git
cd onediff 
git checkout dev_ipa+isntantid
pip install -e .
# onediff_comfy_nodes
cp -r onediff_comfy_nodes path/to/ComfyUI/custom_nodes/
```

**Comment by Gutianpei** [comment]
> > I'm having two issues with this branch when compiling IPAdapter and Animatediff together:
> 
> Fixed the use of AnimateDiff combined with IPA in the same way as before. Please update the code accordingly。
> 
> ### Reference installation
> ```shell
> # onediff
> git clone https://github.com/siliconflow/onediff.git
> cd onediff 
> git checkout dev_ipa+isntantid
> pip install -e .
> # onediff_comfy_nodes
> cp -r onediff_comfy_nodes path/to/ComfyUI/custom_nodes/
> ```

Thanks for the fix! It works well on both of the cases. I'm having a new issue that would appreciate your help:
I'm trying to use `Batch Value Schedule` node to do weight scheduling for IPAdpater, however it will cause oneflow error. I updated https://github.com/siliconflow/onediff/blob/dev_ipa%2Bisntantid/onediff_comfy_nodes/modules/oneflow/infer_compiler_registry/register_comfy/CrossAttentionPatch.py to match the latest IPAdapter and clear the error, however the weight scheduling is still not working. 

I've done some debugging and it looks like the "weight" parameter is correctly passed in and updated in `CrossAttentionPatch.py`, but the "sub_idxs" in `extra_options['ad_params']` is somehow not updated correctly and the value is different with `transformer_options["ad_params"]["sub_idxs"]` in `hijack_sampler.py`.

Any insight would be very helpful! Thanks again for the amazing project!

File: onediff_comfy_nodes/modules/oneflow/infer_compiler_registry/register_comfy/CrossAttentionPatch.py [hunk 1]
```diff
@@ -292,33 +227,68 @@ def forward(self, q, k, v, extra_options):
...
+
     def to(self, *args, **kwargs):
-        # print("Warning: CrossAttentionPatch.to() is called, but it is not implemented yet.")
         return self
     
-    def update(self, patch_kwargs):
-        weight = patch_kwargs.pop("weight")
-        self.weights[0].copy_(torch.Tensor(weight))
+    def bind_model(self, model: torch.nn.Module):
+        self._bind_model = model
```
Comment (ccssu) [inline onediff_comfy_nodes/modules/oneflow/infer_compiler_registry/register_comfy/CrossAttentionPatch.py:247]:
```python
object.__setattr__(self, "_modules", torch_module._modules)
        object.__setattr__(self, "_torch_module", torch_module)
```

File: onediff_comfy_nodes/modules/oneflow/hijack_ipadapter_plus/set_model_patch_replace.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+from ..patch_management import PatchType, create_patch_executor
+
+
+def set_model_patch_replace(org_fn, model, patch_kwargs, key):
+    diff_model = model.model.diffusion_model
+    cache_patch_executor = create_patch_executor(PatchType.CachedCrossAttentionPatch)
+    masks_patch_executor = create_patch_executor(PatchType.CrossAttentionForwardMasksPatch)
+    cache_dict = cache_patch_executor.get_patch(diff_model)
+    cache_key = create_patch_executor(PatchType.UiNodeWithIndexPatch).get_patch(model)
+    to = model.model_options["transformer_options"]
```
Comment (ccssu) [inline onediff_comfy_nodes/modules/oneflow/hijack_ipadapter_plus/set_model_patch_replace.py:16]:
```
op {
    name: "model.input_blocks.1.1.transformer_blocks.0-cast-87"
    device_tag: "cuda"
    scope_symbol_id: 272
    loc: "File \"/data/home/fengwen/quant/ComfyUI/custom_nodes/onediff_comfy_nodes/modules/oneflow/infer_compiler_registry/register_comfy/CrossAttentionPatch.py\", line 197, in forward, source < out_ip = out_ip * weight # I\'m doing this to get the same results as before >; File \"/data/home/fengwen/miniconda3/envs/py310/lib/python3.10/site-packages/oneflow/nn/modules/module.py\", line 225, in __call__, source < res = self.forward(*args, **kwargs) >; ... 57 more"
    user_conf {
      op_type_name: "cast"
      input {
        key: "in"
        value {
          s: "model.input_blocks.1.1.transformer_blocks.0-FreeEagerTensor0-86/out"
        }
      }
      output {
        key: "out"
        value {
          s: "model.input_blocks.1.1.transformer_blocks.0-cast-87/out_0"
        }
      }
      attr {
        key: "dtype"
        value {
          at_data_type: kFloat16
        }
      }
      attr {
        key: "pin_memory"
        value {
          at_bool: false
        }
      }
      input_order: "in"
      output_order: "out"
    }
  }
```

File: onediff_comfy_nodes/_config.py [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+
+def is_disable_oneflow_backend():
+    return _disable_oneflow_backend
+
+
+def is_default_using_nexfort_backend():
+    return _default_backend == "nexfort"
+
+def is_default_using_oneflow_backend():
+    return _default_backend == "oneflow" and not _disable_oneflow_backend
```
Comment (strint) [inline onediff_comfy_nodes/_config.py:22]: 需要 format 一下

File: onediff_comfy_nodes/modules/oneflow/infer_compiler_registry/register_comfy/CrossAttentionPatch.py [hunk 4]
```diff
@@ -292,33 +227,70 @@ def forward(self, q, k, v, extra_options):
...
+
     def to(self, *args, **kwargs):
-        # print("Warning: CrossAttentionPatch.to() is called, but it is not implemented yet.")
         return self
     
-    def update(self, patch_kwargs):
-        weight = patch_kwargs.pop("weight")
-        self.weights[0].copy_(torch.Tensor(weight))
+    def bind_model(self, model: torch.nn.Module):
+        # self._bind_model = model
```
Comment (strint) [inline onediff_comfy_nodes/modules/oneflow/infer_compiler_registry/register_comfy/CrossAttentionPatch.py:247]: 这个可以删了

**Review by strint — APPROVED** [review]

File: onediff_comfy_nodes/modules/oneflow/infer_compiler_registry/register_comfy/CrossAttentionPatch.py (hunk 4 above)
Comment (ccssu) [reply onediff_comfy_nodes/modules/oneflow/infer_compiler_registry/register_comfy/CrossAttentionPatch.py]: done

File: onediff_comfy_nodes/_config.py (hunk 3 above)
Comment (ccssu) [reply onediff_comfy_nodes/_config.py]: done

**Comment by xiecon** [comment]
```
Can't pickle <class 'onediff.infer_compiler.transform.builtin_transform.<class 'register_comfy.CrossAttentionPatch.CrossAttentionPatch'>'>: attribute lookup <class 'register_comfy.CrossAttentionPatch.CrossAttentionPatch'> on onediff.infer_compiler.transform.builtin_transform failed
```
I meet this problem. Also, this problem seems to have caused an increase in the time consumed after enabling onediff in the IPAdapter FaceID node.

**Pull request merged**